File: references/java-sdk/src/main/java/blue/language/sdk/AccessBuilder.java
package blue.language.sdk;

import blue.language.sdk.internal.StepsBuilder;
import blue.language.types.myos.SingleDocumentPermissionGranted;
import blue.language.types.myos.SingleDocumentPermissionRejected;
import blue.language.types.myos.SingleDocumentPermissionRevoked;
import blue.language.types.myos.SubscriptionToSessionFailed;
import blue.language.types.myos.SubscriptionToSessionInitiated;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.function.Consumer;

public final class AccessBuilder<P extends DocBuilder<P>> {

    private final P parent;
    private final String name;
    private Object targetSessionId;
    private String onBehalfOfChannel;
    private boolean read;
    private final List<String> operations = new ArrayList<String>();
    private String statusPath;
    private boolean subscribeAfterGranted;
    private final List<Class<?>> subscriptionEvents = new ArrayList<Class<?>>();
    private boolean subscribeToCreatedSessions;

    private AccessConfig.PermissionTiming permissionTiming = AccessConfig.PermissionTiming.ON_INIT;
    private Class<?> permissionTriggerEventClass;
    private String permissionTriggerDocPath;

    AccessBuilder(P parent, String name) {
        if (parent == null) {
            throw new IllegalArgumentException("parent is required");
        }
        if (name == null || name.trim().isEmpty()) {
            throw new IllegalArgumentException("access name is required");
        }
        this.parent = parent;
        this.name = name.trim();
    }

    public AccessBuilder<P> targetSessionId(Object targetSessionId) {
        this.targetSessionId = targetSessionId;
        return this;
    }

    public AccessBuilder<P> onBehalfOf(String channelKey) {
        this.onBehalfOfChannel = channelKey;
        return this;
    }

    public AccessBuilder<P> read(boolean read) {
        this.read = read;
        return this;
    }

    public AccessBuilder<P> operations(String... operations) {
        if (operations != null) {
            this.operations.addAll(Arrays.asList(operations));
        }
        return this;
    }

    public AccessBuilder<P> statusPath(String statusPath) {
        this.statusPath = statusPath;
        return this;
    }

    public AccessBuilder<P> subscribeAfterGranted() {
        this.subscribeAfterGranted = true;
        return this;
    }

    public AccessBuilder<P> subscriptionEvents(Class<?>... eventTypes) {
        if (eventTypes != null) {
            this.subscriptionEvents.addAll(Arrays.asList(eventTypes));
        }
        return this;
    }

    public AccessBuilder<P> subscribeToCreatedSessions(boolean enabled) {
        this.subscribeToCreatedSessions = enabled;
        return this;
    }

    public AccessBuilder<P> requestPermissionOnInit() {
        this.permissionTiming = AccessConfig.PermissionTiming.ON_INIT;
        this.permissionTriggerEventClass = null;
        this.permissionTriggerDocPath = null;
        return this;
    }

    public AccessBuilder<P> requestPermissionOnEvent(Class<?> eventTypeClass) {
        if (eventTypeClass == null) {
            throw new IllegalArgumentException("eventTypeClass is required");
        }
        this.permissionTiming = AccessConfig.PermissionTiming.ON_EVENT;
        this.permissionTriggerEventClass = eventTypeClass;
        this.permissionTriggerDocPath = null;
        return this;
    }

    public AccessBuilder<P> requestPermissionOnDocChange(String path) {
        if (path == null || path.trim().isEmpty()) {
            throw new IllegalArgumentException("path is required");
        }
        this.permissionTiming = AccessConfig.PermissionTiming.ON_DOC_CHANGE;
        this.permissionTriggerDocPath = path.trim();
        this.permissionTriggerEventClass = null;
        return this;
    }

    public AccessBuilder<P> requestPermissionManually() {
        this.permissionTiming = AccessConfig.PermissionTiming.MANUAL;
        this.permissionTriggerEventClass = null;
        this.permissionTriggerDocPath = null;
        return this;
    }

    public P done() {
        validate();

        String token = DocBuilder.tokenOf(name, "ACCESS");
        String requestId = "REQ_ACCESS_" + token;
        String subscriptionId = "SUB_ACCESS_" + token;

        AccessConfig config = new AccessConfig(
                name,
                token,
                targetSessionId,
                onBehalfOfChannel,
                requestId,
                subscriptionId,
                read,
                operations,
                statusPath,
                subscribeAfterGranted,
                subscriptionEvents,
                subscribeToCreatedSessions,
                permissionTiming,
                permissionTriggerEventClass,
                permissionTriggerDocPath);
        parent.registerAccessConfig(config);

        if (config.statusPath() != null) {
            parent.field(config.statusPath(), "pending");
        }

        if (config.permissionTiming() != AccessConfig.PermissionTiming.MANUAL) {
            String workflowKey = "access" + token + "RequestPermission";
            Consumer<StepsBuilder> permissionStep = steps -> steps.myOs().requestSingleDocPermission(
                    config.onBehalfOfChannel(),
                    config.requestId(),
                    config.targetSessionId(),
                    config.permissionsNode(),
                    config.subscribeToCreatedSessions());
            switch (config.permissionTiming()) {
                case ON_INIT:
                    parent.onInit(workflowKey, permissionStep);
                    break;
                case ON_EVENT:
                    parent.onEvent(workflowKey, config.permissionTriggerEventClass(), permissionStep);
                    break;
                case ON_DOC_CHANGE:
                    parent.onDocChange(workflowKey, config.permissionTriggerDocPath(), permissionStep);
                    break;
                default:
                    break;
            }
        }

        parent.onMyOsResponse("access" + token + "Granted",
                SingleDocumentPermissionGranted.class,
                config.requestId(),
                steps -> {
                    if (config.statusPath() != null) {
                        steps.replaceValue("MarkAccessGranted", config.statusPath(), "granted");
                    }
                    if (config.subscribeAfterGranted()) {
                        Class<?>[] eventTypes = config.subscriptionEvents().toArray(new Class<?>[0]);
                        steps.myOs().subscribeToSession(
                                config.onBehalfOfChannel(),
                                config.targetSessionId(),
                                config.subscriptionId(),
                                eventTypes);
                    }
                });

        parent.onMyOsResponse("access" + token + "Rejected",
                SingleDocumentPermissionRejected.class,
                config.requestId(),
                steps -> {
                    if (config.statusPath() != null) {
                        steps.replaceValue("MarkAccessRejected", config.statusPath(), "rejected");
                    }
                });

        parent.onMyOsResponse("access" + token + "Revoked",
                SingleDocumentPermissionRevoked.class,
                config.requestId(),
                steps -> {
                    if (config.statusPath() != null) {
                        steps.replaceValue("MarkAccessRevoked", config.statusPath(), "revoked");
                    }
                });

        if (config.subscribeAfterGranted()) {
            parent.onSubscriptionUpdate("access" + token + "SubscriptionReady",
                    config.subscriptionId(),
                    SubscriptionToSessionInitiated.class,
                    steps -> {
                        if (config.statusPath() != null) {
                            steps.replaceValue("MarkAccessSubscribed", config.statusPath(), "subscribed");
                        }
                    });
            parent.onEvent("access" + token + "SubscriptionFailed",
                    SubscriptionToSessionFailed.class,
                    steps -> {
                        if (config.statusPath() != null) {
                            steps.replaceValue("MarkAccessSubscriptionFailed", config.statusPath(), "subscription-failed");
                        }
                    });
        }

        return parent;
    }

    private void validate() {
        if (targetSessionId == null) {
            throw new IllegalArgumentException("access('" + name + "'): targetSessionId is required");
        }
        if (onBehalfOfChannel == null || onBehalfOfChannel.trim().isEmpty()) {
            throw new IllegalArgumentException("access('" + name + "'): onBehalfOf is required");
        }
    }
}

File: libs/dsl-sdk/src/lib/interactions/access-builder.ts
import { toTypeAlias, type TypeLike } from '../core/type-alias.js';
import type {
  AccessConfig,
  AccessConfigRegistrationHost,
  InteractionPermissionTiming,
} from './types.js';

function requireText(value: string | undefined, field: string): string {
  const normalized = value?.trim();
  if (!normalized) {
    throw new Error(`${field} is required`);
  }
  return normalized;
}

function token(name: string): string {
  const normalized = name
    .replace(/[^A-Za-z0-9]+/gu, '_')
    .replace(/^_+|_+$/gu, '')
    .toUpperCase();
  return normalized.length > 0 ? normalized : 'ACCESS';
}

type StepsCustomizer = (steps: {
  myOs: (adminChannelKey?: string) => {
    requestSingleDocPermission: (
      onBehalfOf: string,
      requestId: string,
      targetSessionId: string,
      permissions: unknown,
    ) => unknown;
    subscribeToSession: (
      targetSessionId: string,
      subscriptionId: string,
      ...eventTypes: string[]
    ) => unknown;
  };
  replaceValue: (name: string, path: string, value: unknown) => unknown;
}) => void;

type DocBuilderLike = {
  field(path: string, value: unknown): unknown;
  onInit(workflowKey: string, customizer: StepsCustomizer): unknown;
  onEvent(
    workflowKey: string,
    eventType: string,
    customizer: StepsCustomizer,
  ): unknown;
  onDocChange(
    workflowKey: string,
    path: string,
    customizer: StepsCustomizer,
  ): unknown;
  onMyOsResponse(
    workflowKey: string,
    responseType: string,
    requestIdOrMatcher: string | Record<string, unknown>,
    customizer: StepsCustomizer,
  ): unknown;
  onSubscriptionUpdate(
    workflowKey: string,
    subscriptionId: string,
    updateType: string,
    customizer: StepsCustomizer,
  ): unknown;
};

export class AccessBuilder<P> {
  private onBehalfOfValue: string | undefined;
  private targetSessionIdValue: string | undefined;
  private requestIdValue: string | undefined;
  private subscriptionIdValue: string | undefined;
  private readValue = true;
  private readonly operationsValue: string[] = [];
  private statusPathValue: string | undefined;
  private subscribeAfterGrantedValue = false;
  private readonly subscriptionEventsValue: string[] = [];
  private subscribeToCreatedSessionsValue = false;
  private permissionTimingValue: InteractionPermissionTiming = 'onInit';
  private permissionTriggerEventTypeValue: string | undefined;
  private permissionTriggerDocPathValue: string | undefined;

  constructor(
    private readonly parent: AccessConfigRegistrationHost<P>,
    private readonly name: string,
  ) {}

  onBehalfOf(channelKey: string): this {
    this.onBehalfOfValue = requireText(channelKey, 'onBehalfOf');
    return this;
  }

  permissionFrom(channelKey: string): this {
    this.onBehalfOf(channelKey);
    return this;
  }

  targetSessionId(sessionId: string): this {
    this.targetSessionIdValue = requireText(sessionId, 'targetSessionId');
    return this;
  }

  requestId(requestId: string): this {
    this.requestIdValue = requireText(requestId, 'requestId');
    return this;
  }

  subscriptionId(subscriptionId: string): this {
    this.subscriptionIdValue = requireText(subscriptionId, 'subscriptionId');
    return this;
  }

  read(read: boolean): this {
    if (read === false) {
      throw new Error(
        'access(...).read(false) is not supported; MyOS single-document permissions require read=true',
      );
    }
    this.readValue = read;
    return this;
  }

  operations(...operations: string[]): this {
    for (const operation of operations) {
      const normalized = operation?.trim();
      if (!normalized) {
        continue;
      }
      this.operationsValue.push(normalized);
    }
    return this;
  }

  statusPath(statusPath: string): this {
    this.statusPathValue = requireText(statusPath, 'statusPath');
    return this;
  }

  subscribeAfterGranted(): this {
    this.subscribeAfterGrantedValue = true;
    return this;
  }

  subscriptionEvents(...eventTypes: TypeLike[]): this {
    this.subscriptionEventsValue.length = 0;
    this.subscriptionEventsValue.push(
      ...eventTypes.map((eventType) => toTypeAlias(eventType)),
    );
    return this;
  }

  subscribeToCreatedSessions(enabled = true): this {
    if (enabled) {
      throw new Error(
        'access(...).subscribeToCreatedSessions(true) is not supported on the current public runtime',
      );
    }
    this.subscribeToCreatedSessionsValue = enabled;
    return this;
  }

  requestPermissionOnInit(): this {
    this.permissionTimingValue = 'onInit';
    this.permissionTriggerEventTypeValue = undefined;
    this.permissionTriggerDocPathValue = undefined;
    return this;
  }

  requestPermissionOnEvent(eventType: TypeLike): this {
    this.permissionTimingValue = 'onEvent';
    this.permissionTriggerEventTypeValue = toTypeAlias(eventType);
    this.permissionTriggerDocPathValue = undefined;
    return this;
  }

  requestPermissionOnDocChange(path: string): this {
    this.permissionTimingValue = 'onDocChange';
    this.permissionTriggerDocPathValue = requireText(path, 'permission path');
    this.permissionTriggerEventTypeValue = undefined;
    return this;
  }

  requestPermissionManually(): this {
    this.permissionTimingValue = 'manual';
    this.permissionTriggerDocPathValue = undefined;
    this.permissionTriggerEventTypeValue = undefined;
    return this;
  }

  done(): P {
    const baseToken = token(this.name);
    const config: AccessConfig = {
      name: this.name,
      token: baseToken,
      targetSessionId: requireText(
        this.targetSessionIdValue,
        'targetSessionId',
      ),
      onBehalfOf: requireText(this.onBehalfOfValue, 'onBehalfOf'),
      requestId: this.requestIdValue ?? `REQ_ACCESS_${baseToken}`,
      subscriptionId: this.subscriptionIdValue ?? `SUB_ACCESS_${baseToken}`,
      read: this.readValue,
      operations: [...new Set(this.operationsValue)],
      ...(this.statusPathValue ? { statusPath: this.statusPathValue } : {}),
      subscribeAfterGranted: this.subscribeAfterGrantedValue,
      subscriptionEvents: [...this.subscriptionEventsValue],
      subscribeToCreatedSessions: this.subscribeToCreatedSessionsValue,
      permissionTiming: this.permissionTimingValue,
      ...(this.permissionTriggerEventTypeValue
        ? { permissionTriggerEventType: this.permissionTriggerEventTypeValue }
        : {}),
      ...(this.permissionTriggerDocPathValue
        ? { permissionTriggerDocPath: this.permissionTriggerDocPathValue }
        : {}),
      permissionFrom: this.onBehalfOfValue,
    };
    const registered = this.parent.registerAccessConfig(config);
    this.applyAutoWiring(config, registered as unknown as DocBuilderLike);
    return registered;
  }

  private applyAutoWiring(config: AccessConfig, parent: DocBuilderLike): void {
    if (config.statusPath) {
      parent.field(config.statusPath, 'pending');
    }

    const workflowPrefix = `access${config.token}`;
    const requestPermissionWorkflow: StepsCustomizer = (steps) => {
      const permissions: Record<string, unknown> = {};
      if (config.read) {
        permissions.read = true;
      }
      if (config.operations.length > 0) {
        permissions.singleOps = [...config.operations];
      }
      steps
        .myOs()
        .requestSingleDocPermission(
          config.onBehalfOf,
          config.requestId,
          config.targetSessionId,
          permissions,
        );
    };

    if (config.permissionTiming === 'onInit') {
      parent.onInit(
        `${workflowPrefix}RequestPermission`,
        requestPermissionWorkflow,
      );
    } else if (config.permissionTiming === 'onEvent') {
      parent.onEvent(
        `${workflowPrefix}RequestPermission`,
        config.permissionTriggerEventType ?? 'Conversation/Event',
        requestPermissionWorkflow,
      );
    } else if (config.permissionTiming === 'onDocChange') {
      parent.onDocChange(
        `${workflowPrefix}RequestPermission`,
        config.permissionTriggerDocPath ?? '/',
        requestPermissionWorkflow,
      );
    }

    parent.onMyOsResponse(
      `${workflowPrefix}Granted`,
      'MyOS/Single Document Permission Granted',
      config.requestId,
      (steps) => {
        if (config.statusPath) {
          steps.replaceValue('MarkAccessGranted', config.statusPath, 'granted');
        }
        if (config.subscribeAfterGranted) {
          steps
            .myOs()
            .subscribeToSession(
              config.targetSessionId,
              config.subscriptionId,
              ...config.subscriptionEvents,
            );
        }
      },
    );

    parent.onMyOsResponse(
      `${workflowPrefix}Rejected`,
      'MyOS/Single Document Permission Rejected',
      config.requestId,
      (steps) => {
        if (config.statusPath) {
          steps.replaceValue(
            'MarkAccessRejected',
            config.statusPath,
            'rejected',
          );
        }
      },
    );

    parent.onMyOsResponse(
      `${workflowPrefix}Revoked`,
      'MyOS/Single Document Permission Revoked',
      config.requestId,
      (steps) => {
        if (config.statusPath) {
          steps.replaceValue('MarkAccessRevoked', config.statusPath, 'revoked');
        }
      },
    );

    if (!config.subscribeAfterGranted) {
      return;
    }

    parent.onMyOsResponse(
      `${workflowPrefix}SubscriptionReady`,
      'MyOS/Subscription to Session Initiated',
      {
        subscriptionId: config.subscriptionId,
      },
      (steps) => {
        if (config.statusPath) {
          steps.replaceValue(
            'MarkAccessSubscribed',
            config.statusPath,
            'subscribed',
          );
        }
      },
    );

    parent.onMyOsResponse(
      `${workflowPrefix}SubscriptionFailed`,
      'MyOS/Subscription to Session Failed',
      {
        subscriptionId: config.subscriptionId,
      },
      (steps) => {
        if (config.statusPath) {
          steps.replaceValue(
            'MarkAccessSubscriptionFailed',
            config.statusPath,
            'subscription-failed',
          );
        }
      },
    );
  }
}
